README.md
# The I-Will System &mdash; a.k.a. "promises.to"

NOTE: The spec here is evolving. See also the comments in server.js

Problem:
Alice casually says to Bob 
"I'll let you know if I can make it to the meeting"
but then forgets to follow through.

Solution:
Any time Alice makes any 
"<a href="http://blog.beeminder.com/will">I will</a>" 
statement like that, she types a URL like so:

`alice.promises.to/let_bob_know_re_meeting/by/tomorrow_5pm`

As in, she literally types that directly to Bob, manually, when she's making the promise to him.
When Alice or Bob click that URL a promise is created in the promises.to app and a calendar
entry is added to Alice's calendar and a datapoint is sent to Beeminder.

## What happens on the first click of a promise URL

1. It should completely ignore whether the user is logged in. If you
go to alice.promises.to/foo/by/xyz then it says "click to confirm that
alice promises to 'foo' by [date-picker]". The datepicker uses the
"/by/xyz" from the URL as the default but if it parsed it wrong you
have a chance to change it.

2. So the promise isn't actually saved until that button is pressed
but the wording on the page encourages the recipient of the promise to
click it. Or the promiser (call her Alice) can do it -- it doesn't
matter who confirms the creation of the promise.

3. All subsequent visits to the URL just say "alice has promised to
'foo' by [date]" and maybe also a nice countdown timer to the deadline
to give a feeling of a ticking clock. Also it should show Alice's
reliability statistics. Maybe it shows Alice's full list of promises.
Same as if you visit alice.promises.to with no path in the URL.

4. And the other cool part: when a deadline passes without being
marked fulfilled then the reliability score decreases in real time
according to the late penalty function. (See "Computing Statistics" below.)

5. So the only difference for the logged in user is the ability to
mark promises fulfilled. (Post-MVP maybe there's a way for the
promisee (ie, person the promise was made to) to be the one to mark it
fulfilled. And for the MVP maybe everything can be wide open -- we're
not going to have any abuse until there are, realistically, thousands
of users or more.)


----

Everything below is in flux

----

> *Nerd note: Creating an object in a database in response to a GET request is
> pretty unkosher. 
> We're going to do it anyway because of how elegently it reduces the friction for the user.
> Otherwise Alice has to type the promise to Bob and then separately use some UI to add the
> promise to the system.*

The Beeminder datapoint gets the specified deadline as the date (even though it's in the future) 
and a zero as the datapoint value.
The comment should include the deadline time of day and when the promise was first created.

> *For later: 
> Bee doesn't like the idea of events getting created on her calendar whenever random URLs
> get surfed to.
> Probably it will be fine and is good enough for the MVP but later we'll consider an 
> enhancement like so:
> Leave the deadline off of the URL and the promise is created but with no calendar entry.
> When the user is logged in, all such promises are shown and the user can explicitly add a
> deadline.
> Since the user created the promise with a highly greppable string in chat or email it's easy
> for them to look up the deadline they mentioned when they created the promise.
> At that point the calendar entry is created but it took the logged in user doing it.
> For sufficient paranoia an account-level setting could allow or disallow deadlines specified
> as part of the URL.
> (Also some initial feedback suggests that URLs like
> `alice.promises.to/do_the_thing/by/8am_tue`  
> may be too intimidating for nontechnical folks but just
> `alice.promises.to/do_the_thing`  
> is ok.)*

After the promise is created, the app redirects to the main `alice.promises.to` page.

## Uniqueness of Promise Names

The identifier for a promise is the name the user typed in the URL, 
the `let_bob_know_re_meeting` part, 
concatenated with the unixtime (in seconds) of the deadline.
The deadline is specified in the URL by something like "10pm_tue"
so if you make another promise of `let_bob_know_re_meeting` 
next week that will be created as a distinct promise and everything will work fine.
Repeat hits on that URL *this* week will resolve to the same promise identifier 
and be idempotent.

## Account Settings

1. Username, used as a subdomain for the URL
2. Timezone (needed to parse the deadlines)

Later:

1. Pronoun (default "they/them/their/theirs")
2. Display name, e.g., "Alice" as opposed to username "alice"

## What Non-Logged-In Users See

Minimally "alice.promises.to" should show this:
  
> If you're here then alice must've promised you they'd do something.
> They've made N such promises so far with a success rate of X%!

It could also show her Beeminder graph.

## UI For Marking Promises Fulfilled


If Alice is logged in, the app lists her existing promises and lets her choose one of 
{not fulfilled yet, fulfilled on time, fulfilled partially or late} 
for each of them. 
Each defaults to "not fulfilled yet". 
If fulfilled partially, Alice specifies what percent fulfilled it was and/or when it was fulfilled.

Whenever anything about the promise changes it needs to be automatically mirrored in Beeminder.

We may also need an option to delete promises or mark them as voided.
We're not sure what should happen on Beeminder with a voided promise so we're just not going to
implement promise-voiding until there's demand for such a feature.

## Computing Statistics

The only statistics we'll care about initially are the number of promises Alice has made 
and her reliability percentage.
And pending vs past promises.

Definitions:

* `fill` &mdash; a promise's fulfilled fraction, between 0 to 1, default 0
* `t` &mdash; current unixtime
* `tfin` &mdash; promise's deadline

A promise's fulfilled fraction, `fill`, is 1 if fulfilled on time, or the specified percentage. 
If there's a fulfilled time specified then `fill` = the specified percentage (or 1 if not specified) times `credit(tfin-t)` where the credit function maps a number of seconds to how much credit you get if you're that much late:

    credit(t) = 1    if t<60              # seconds late
    credit(t) = .999 if t<3600            # minutes late
    credit(t) = .99  if t<86400           # hours late
    credit(t) = .9   if t<86400*7         # days late
    credit(t) = .5   if t<86400*365.25/12 # weeks late
    credit(t) = .1   if t<86400*365.25    # months late
    credit(t) = .01  if t<86400*365.25*10 # years late
    credit(t) = 0    otherwise            # more than a decade late

> *Implementation note: Dreev has a version of that function ready that interpolates continuously.
> We'll want that version because it will be super cool to see the reliability percentage tick
> down in real time when one of Alice's deadlines passes.
> I know React makes that kind of thing easy.*

> *An alternative simple late penalty function I've used before in another context is 
> `1-(t/(86400*7))^4` 
> which gives you a couple days of very little penalty and then your credit plummets 
> till you get zero credit for being more than a week late.*

Finally, for the statistics, iterate through the promises, `p`, like so:

    if p is marked "not fulfilled yet" then
      if t<tfin then                # unfulfilled future promises don't
        waiting++                   #  count for or against you.
      else 
        numerator += credit(t-tfin) # optimistically assume you're just
        denominator++               #  about to fulfill the promise.
      end
    else                            # p is marked fulfilled or partially
      numerator += fill             #  fulfilled, whether or not it's
      denominator++                 #  past the deadline.
    end

That's it! 
Now you can report that the user has made 
{`denominator+waiting`} promises 
(of which {`waiting`} are still in the future) 
and has a reliability of 
{`numerator/denominator*100`}%.


# Credits

Daniel Reeves wrote a blog post about the idea.
Sergii Kalinchuk got the "promises.to" domain.
Marcin Borkowski had the idea for URLs-as-UI for creating promises.
Chris Butler wrote some of the initial code.

<br>
<br>
<br>


## For Later: Calendar as UI

Maybe it's not too much friction to just manually add entries to your calendar as the primary interface for adding promises. 
There are various ways to do that with very low friction already. 
But then that needs to automatically trigger promises.to to capture each calendar entry.
(I'm doing that now with IFTTT to send promises to Beeminder.)

And maybe it's fine for *every* calendar entry to get automatically added. 
Some of them won't be promises but that's fine -- you can just mark them as 
non-promises or delete them and they won't count. 
If they are promises then you need to manually mark them as fulfilled or not. 
Beeminder (plus the embarrassment of having your reliability percentage drop 
when a deadline passes) should suffice to make sure you remember to do that.

## For Later: Security and Privacy

Alice's friends can troll her by making up URLs like 
alice.promises.to/kick_a_puppy 
but that's not a huge concern. 
Alice, when logged in, has to approve promises to be public. 
So the prankster will see a page that says Alice promises to kick a puppy 
but no one else will.

In the MVP we can skip the approval UI and worry about abuse like that the first 
time it's a problem, 
which I predict will be after promises.to is a million dollar company.

## Automatically Creating Calendar Entries

It's pretty critical that the promises end up on your calendar.
That could be done semi-manually by creating links like described here: <https://stackoverflow.com/questions/10488831/link-to-add-to-google-calendar>  
No Calendar API needed that way -- just construct the link and if the user is 
logged in to Google it will create the calendar entry when the click it.

## GET vs POST

If create-on-GET is just too hideous then the GET on alice.promises.to/foo
could show a page like:

> Did alice just promise to foo?
>
> [big button that says "YES"]

But I think anyone should be able to click yes, no authentication required.

And I really want to try the outrageous create-on-GET version first because I 
think it might be a feature that every yourname.promises.to URL you type gets 
almost automatically logged as a promise.
And I suspect that spiders and such generating URLs you didn't type will be a non-issue.


## More on Duplicates

When you visit alice.promises.to/foo it shows you,
in addition to the current promise that hitting that link created,
all the past promises named foo and when they were due, etc.

Also I think the deadline should default to 5pm the following business day.
Maybe in the initial implementation that's all that's supported.
No parsing the messy "/by/" stuff in the URLs and no UI for changing deadlines.

## More on Beeminder Integration

Mine is just a do-more goal to fulfill like 8 promises per week. 
I create a datapoint for each promise (via IFTTT from Google Calendar) 
when I promise it, and then change the datapoint to a 1 when I fulfill it 
(or something less than 1 if I fulfill it late).

So Beeminder is not enforcing a success rate, just an absolute number of successes.

## Getting something dogfoodable as quickly as possible

1. no parsing deadlines, just always assume 5pm the next business day
2. no UI for anything, the page just lists all promises no matter what
3. do need a way to mark promises complete (tack on a query param to do it?)
4. no privacy or security features, everything is public
5. no calendar API, just construct a link the user can click to create the calendar event

## Other domain name ideas

* dreev.es/will/ (for anyone who has a domain for their own name)
* alice.promises.to/ (sergii grabbed this one)
* alice.commits.to/ (dreev grabbed this one)
* alice.willdefinite.ly/ (kinda awkward)
* alice.willveri.ly/ (too cutesy?)
* alice.willprobab.ly/ (emphasizes the reliability percentage)
* alice.willresolute.ly

# Changelog

View the [project history](https://glitch.com/edit/#!/iwill?path=CHANGELOG.md:1:0)



# Next tasks

1. finish the parseProm function based on the new simple parsing rules
2. add all the fields to the database
3. parse incoming promises so all the fields are stored
4. anything not parseable by the simple rules yields an error message that the user sees when clicking on the URL
5. interface to mark promises completed
6. real-time reliability score


history.md
History
---

##### 09-23-17 - Chris
 - Don't create duplicate promises when the same link is clicked
 - Add a 'promise' route that answers when a promise already exists

server.js
// --------------------------------- 80chars ---------------------------------->
var express = require('express')
var Sequelize = require('sequelize')
var app = express()
var chrono = require('chrono-node') // Sugar.js parses some things better
var Mailgun = require('mailgun-js')
var moment = require('moment')

// DATABASE FIELDS:
//   urtx -- urtext, ie, full original text the user typed to create the promise
//   user -- who's making the promise
//   what -- what's being promised
//   whom -- to whom are you promising
//   tini -- unixtime that the promise was made
//   tdue -- unixtime that the promise is due
//   tmzn -- timezone assumed for parsing the deadline
//   tfin -- unixtime that the promise was fulfilled
//   fill -- fraction fulfilled, default 0
//   void -- true if the promise became unfulfillable or moot
//   clix -- number of clicks a promise has gotten
// For example:
//   urtx = "bob.promises.to/foo_the_bar/by/noon_tomorrow"
//   user = "bob"
//   what = "foo the bar"
//   whom = null
//   tini = [unixtime of first GET request of the promise's URL]
//   tdue = [what "noon tomorrow" parsed to at time tini]
//   tmzn = "America/Los_Angeles"
//   tfin = [unixtime that the promise was fulfilled]
//   fill = 0
//   void = false

// Other ideas for fields: 
// * information about the client that originally created the promise
// * whether the promise was created by the actual user (if they were logged in 
//   and were the first to click on it) or by another logged-in user or by 
//   someone not logged in
// * conf: maybe we create the promise whether or not anyone clicks the button to confirm
//   it, in which case we store when it's actually been confirmed.

// LATEST SPEC:
// The "/by/..." part is an optional power-user thing, not part of the MVP.
// Instead, the first time you go to alice.promises.to/foo it gives you the 
// option to choose a deadline, defaulting to now+7days (or to what the "/by/" 
// part parsed to, if that's easy, but just make it a suggestion for the user, 
// just like now+7days is if they don't specify a "/by/"). 
// For every subsequent GET of alice.promises.to/foo it shows the deadline, with
// no way to change it.

// IDEAS FOR LATER:
// Simpler date parsing:
//   Maybe the date has to be YYYY-MM-DD_HH:MM
//   Example: bob.promises.to/foo_the_bar/by/2017-09-13_12:00
// Magic spaces:
//   Whichever non-alphanumeric character is most common in the urtext, that's 
//   what's assumed to be a space and is replaced with spaces before parsing.
// Less magical version:
//   A non-alphanumeric character must follow "alice.promises.to/" and that
//   character is taken as the ersatz space. Eg:
//   alice.promises.to/_start_her_promises_with_underscores
// Flexibility on the '/by/' part:
//   Requiring the string '/by/' to appear in the promise URL means no ambiguity
//   about where to start parsing the deadline. But the Chrono parsing library 
//   actually does great taking the whole string like "foo the bar by noon 
//   tomorrow" and figuring out the time. We could also just take the last
//   occurrence of "by" and parse everything after it as the deadline.

app.use(express.static('public'))

function mailself(subj, body) {
  var mailgun = new Mailgun({apiKey: process.env.MAILGUN_KEY, 
                             domain: process.env.MAILGUN_DOMAIN})
  var data = {
    from:    process.env.MAILGUN_FROM,
    to:      process.env.MY_EMAIL,
    subject: subj,
    text:    body,
  }
  mailgun.messages().send(data, (err, body) => {})
}

// Take urtext and return a promise data structure
function parseProm(urtx) {
  let m = urtx.match(/^([^\.]+)\.(promises|commits)\.to\/(.*)$/)
  // TODO: handle the case that no user is given, just "promises.to/foo"
  let user = m ? m[1] : ''
  let domain = m ? m[2] : ''
  let path = m ? m[3] : ''
  let what = ''
  let tdue = Date.now() + 7*24*60*60*1000
  if (path.length > 0) {
    let parsedPath = path.match(/^(.*?)(\/by\/|$)(.*)$/)
    what = parsedPath[1] //.replace(/_/g, ' ')
    //tdue = parsedPath.length > 1 ? 
    //       chrono.parseDate(parsedPath[3].replace(/_/g, ' ')) * 1 : 
    //       Date.now() + 7*24*60*60*1000
    //tdue = tdue ? ( ( tdue < Date.now() && Date.now() - tdue < 24*60*60*1000 ) ?
    //                tdue + 24*60*60*1000 : tdue )
    //            : Date.now() + 7*24*60*60*1000
    tdue = parsedPath[3]
  }
  // TODO: path is now everything after ".promises.to" so now parse out the 
  //       "what" (thing being promised) and the "tdue" (deadline)
  console.log('parseProm', { user, path, domain, what, tdue, urtx })
  return { user, path, domain, what, tdue, urtx, tini: Date.now() }
}

// Compute the credit you get for being t seconds late
function credit(t) {
  // see latepenalty.js
  return 1
}

// Initial promise list hardcoded so don't have to worry about blowing db away
let promises = [
  "alice.promises.to/File_the_TPS_report/by/noon_mon",
  "bob.promises.to/Send_vacation_photos/by/saturday",
  "carol.promises.to/Call_the_dentist/by/12am",
  "dreev.promises.to/show_sergii_this_doing_something_useful/by/3am", // 9-10, done
  "dreev.promises.to/show_bee_open_questions/by/10:30pm", // 9-10, done
  "braden.promises.to/help_with_javascript/by/5pm", // 9-11
  "dreev.promises.to/show_brennan_intro/by/10:50pm", // 9-11, done
  "dreev.promises.to/queue_tweet_etc_for_blog_post/by/2pm", // 9-11, done
  "dreev.promises.to/order_indian/by/5:30pm", // 9-12, done
  "dreev.commits.to/ask_bee_re_sunday_parkways", // 9-13, done
  "dreev.commits.to/follow_up_with_everyone_re_guest_blogging", // 9-13
  "josh.commits.to/get_realisies_running/by/september_30", // 9-13
  "braden.commits.to/outline_bite_counting_post/by/Sunday_11pm", // 9-13
  "dreev.commits.to/report_glitch_bug", // 9-14
  "dreev.commits.to/googledocify_intern_form/by/12pm", // 9-14
  "dreev.promises.to/test_penalty_func/by/saturday_noon", // 9-14
  "sergii.promises.to/work_on_points_2_to_4/by/next_monday", // 9-14
  "dreev.commits.to/meta_debrief_kim/by/midnight_friday", // 9-14
  "dreev.commits.to/ping_mirabai/by/sep_22", // 9-14
  "dreev.commits.to/send_ms_gross_email/by/sunday_11am", // 9-14
  "dreev.promises.to/ping_oli/by/nov_16", // 9-15
  "dreev.promises.to/ping_byorgey/by/dec_20", // 9-15
  "dreev.promises.to/labelzero/by/1am", // 9-15
  "chris.promises.to/follow_up_on_that_support_thread", // 9-18
  "bee.commits.to/new-family-photo-to-yoko/by/tomorrow-night", // 9-18, done early
  "bee.commits.to/rest-of-paperwork-to-yoko-before-the-gym-tomorrow", // 9-18, 2 hours late
  "dreev.promises.to/debug_capitalization/by/tomorrow", // 9-18, done 23.4 hours late
  "bee.commits.to/email-sleep-as-android-for-specifics-about-sleep-length-measurement", // 9-19
  "jessica.commits.to/let_bob_know_re_meeting/by/tomorrow_5pm", // 9-19
  "samuel.commits.to/finish_ch_23_problem_set/by/today_6_pm", // 9-19
  "dreev.promises.to/intern_disc/by/wed", // 9-19, done early
  "kim.promises.to/email_argue_summary/by/fri", // 9-19
  "dreev.commits.to/clear_bogus_commit_for_nick/by/5_minutes_from_now", // 9-19, done early
  "kim.promises.to/reflect_Respond_anxiety_spiral_Assist/by/sat", // 9-19
  "dreev.commits.to/schedule-k_tax_thing/by/5pm", // 9-19, done
  "dreev.promises.to/set_up_mom_on_wordfeud/by/tomorrow", // 9-20, done/void
  "dreev.promises.to/reply_to_use_case_email/by/midnight", // 9-20, done early
  "kim.promises.to/consider_all_weekend_social_options/by/fri/10pm", // 9-20
  "kim.promises.to/submit_conclusion_email/by/10/16", // 9-21
  "kim.promises.to/complete_type_1_essay/by/10/1", // 9-21
  "kim.promises.to/inform_eve_plans/by/2", // 9-21
  "kim.promises.to/email_cynthia_charlap_tickets/by/9/22", // 9-21
  "kim.promises.to/schedule_attorney_meet/by/10/1", // 9-21
  "kim.promises.to/john_PAM_photos_ask/by/10/1", // 9-21
  "kim.promises.to/Nike_p5_email/by/10/1", // 9-21
  "kim.promises.to/intro_martin_marna/by/10/5", // 9-21
  "kim.promises.to/sbonomic_jitsjudgelist_doc/by/10/15", // 9-21
  "kim.promises.to/consider_all_weekend_social_options/by/fri/10pm", // 9-21
  "dreev.commits.to/dedup_the_iwills/by/nov", // 9-22
  "byorgey.promises.to/email_cut_property_summary/by/7pm", // 9-22
  "dreev.commits.to/copy_editing_pass_on_tao2/by/tmw_2pm", // 9-22, done
  "chris.promises.to/open_a_smoothie_shop/by/December", // 9-25
  "pierre.promises.to/water_the_office_plant/by/Friday", // 9-25
  "dreev.promises.to/hanna_reply/by/tonight", // 9-26, done
  "bee.promises.to/read-hannas-emails", // 9-26
  "bee.promises.to/reping-one-with-heart", // 9-26
  "bee.promises.to/fill-out-metromile-feedback", // 9-26
  "bee.promises.to/schedule-planning-with-cantor/by/friday-night", // 9-26
  "dreev.promises.to/ask_jana_about_blood_testers/by/dec_1", // 9-26
  "dreev.commits.to/take_copyediting_pass_on_micheles_draft/by/saturday", // 9/27,
  "stephen.promises.to/decide_upon_late_october_trip/by/saturday", // 9-27
  "roy.promises.to/sign_up_for_app", // 9-27
  "jennyli.promises.to/water_the_plants/by/sunday"  // 9-27
]

let users = [ 
  "alice", "bob", "carol", // testing
  "dreev", "sergii", "kim", "bee", "braden", // initial co-conspirators
  "byorgey", "nick", "josh", "dehowell", "caillu", "mbork", "roy", "jennyli", // daily beemail
  "samuel", "cole", "jessica", "steven", // weekly beemail
  "chris", "stephen", // contributors
  "pierre", // invitees
]

/* 
let test
//test = chrono.parseDate("foo the bar by sep 15th noon")
test = chrono.parseDate("Call_the_dentist by 2017-10-15 12pm")
console.log(`DEBUG: ${test} (unixtime = ${test.getTime()/1000}`)
console.log("DEBUG: ", JSON.stringify(parseProm(promises[2])))
*/

var Promise

// set up a new database using database credentials set in .env
var sequelize = new Sequelize('database', process.env.DB_USER, 
                                          process.env.DB_PASS, {
  host: '0.0.0.0',
  dialect: 'sqlite',
  pool: {
    max: 5,
    min: 0,
    idle: 10000,
  },
  // Security note: db is saved to `.data/database.sqlite` in local filesystem.
  // Nothing in `.data` directory gets copied if someone remixes the project.
  storage: '.data/database.sqlite'
});

// authenticate with the database
sequelize.authenticate()
  .then(function(err) {
    console.log('Connection has been established successfully.')
    // define a new table 'promises'
    Promise = sequelize.define('promises', {
      urtx: { type: Sequelize.STRING  }, // urtext, including whole URL
      user: { type: Sequelize.STRING  }, // who's making the promise
      what: { type: Sequelize.STRING  }, // what's being promised
      //whom: { type: Sequelize.STRING  }, // to whom are you promising
      tini: { type: Sequelize.INTEGER }, // unixtime that promise was made
      tdue: { type: Sequelize.STRING }, // unixtime that the promise is due
      //tmzn: { type: Sequelize.STRING  }, // timezone
      //wtdid: { type: Sequelize.INTEGER }, // unixtime that the promise was fulfilled
      //fill: { type: Sequelize.FLOAT   }, // fraction fulfilled
      //void: { type: Sequelize.BOOLEAN }, // whether promise was voided
      // text:   { type: Sequelize.STRING }, // this was just for testing
    });    
    //setup()
  })
  .catch(function (err) {
    console.log('Unable to connect to the database: ', err)
  })

// populate table with hardcoded promises above
function setup() { 
  Promise.sync({force: true}) // 'force: true' just drops the table if it exists
    .then(function(){         // and creates a new one!
      // Add the default promises to the database
      for (var i = 0; i < promises.length; i++) {
        Promise.create(parseProm(promises[i]))
      }
    })
}

app.get('/promise/:udp/:urtx', function(req, resp) {
  let urtx = req.originalUrl.substr(9)
  Promise.findOne({ where: {urtx}})
    .then(function(promise) {
      console.log('single promise', urtx, promise)
    // resp.write(promise)
    resp.json(promise)
  })

})

app.get('/', (req, resp) => { 
  resp.sendFile(__dirname + '/public/index.html') 
})

app.get('/promises', function(req, resp) {
  var dbPromises = {}
  Promise.findAll({
    order: sequelize.literal('tini DESC')
  }).then(function(promises) {
    console.log('a;; promises', promises)
    promises.forEach(function(promise) {
      dbPromises[promise.user] = dbPromises[promise.user] || []
      dbPromises[promise.user].push(promise) // adds their info to dbPromises value
    });
    resp.send(dbPromises) // sends dbPromises back to the page
  })
})

app.get('/promises/:user', function(req, resp) {
  var dbPromises = [];
  Promise.findAll({
   where: {
     user: req.params.user
   },
  }).then(function(promises) {
    console.log('user promises', promises);
    promises.forEach(function(promise) {
      dbPromises.push(promise) // adds their info to dbPromises value
    });
    resp.send(dbPromises) // sends dbPromises back to the page
  })
})

app.get('/promises/remove/:urtx', (req, resp) => {
  Promise.findAll({
   where: {
     user: req.params.user
   },
  })
})

// creates a new entry in the promises table with the submitted values
app.post('/promises', (req, resp) => {
  console.log('promise req', req, resp);
  Promise.create({ text: req.query.promise })
  resp.sendStatus(200)
})

app.get('/reset', (req, resp) => {
  setup()
  resp.redirect('/')
})

// removes all entries from the promises table
app.get('/clear', (req, resp) => {
  Promise.destroy({where: {}})
  resp.redirect('/')
})

/*
Little picture: Why does this error handling code not run when I surf to 
a bad URL like iwill.glitch.me/foo%bar

Bigger picture: I'm making an app that needs to process any URL the user may 
throw at it. I need the server to get the exact URL the way the user sees it in 
the browser. That's going to be especially tricky for things like '#' characters
but right now I'm just trying to figure out the case of arbitrary '%' characters
that may not correspond to proper %-encodings.

I'm thinking like a catchall error-handling route that no matter what weird
encoding or whatever error is thrown I can still have the code here do something
with the URL as the user typed it.

For now we're just dropping all this and having the app give the user an error
if there are any weird characters in the URL.

app.use(function(req, resp, next) {
  var err = null
  try {
    console.log("TRYING", req.path, req.url)
    decodeURIComponent(req.path)
  } catch(e) {
    err = e
  }
  if (err) {
    console.log("CAUGHT ERR:", err, req.url)
    return resp.redirect('/')
  }
  next();

  //console.log("DEBUG USE1:", req.originalUrl)
  //resp.redirect('/')
})
*/

app.all('*', (req, resp) => {
  // The server at promises.to passes along the full URL the way the user typed
  // it, so when the user hits "bob.promises.to/foo" the Glitch app is called
  // with "iwill.glitch.me/bob.promises.to/foo" and req.originalUrl is
  // "/bob.promises.to/foo" (originalUrl doesn't include the hostname)
  let urtx = req.originalUrl.substr(1) // get rid of the initial slash
  // urtx is now, eg, "bob.promises.to/foo_the_bar/by/9am"
  
  // TODO: don't try to parse a promise if no subdomain, ie, no user given
  // TODOLATER: how to handle "bob.promises.to/foo%bar"
  // TODOLATER: need to url-decode to handle "bob.promises.to/foo^bar"
  // TODOLATER: handle bob.promises.to/do_item_#2/by/9am
  //           https://stackoverflow.com/questions/18796421/capture-anchor-links
  
  console.log("DEBUG urtx:", urtx)
  // Parse the urtext into user, what, whom, tini, tdue, etc
  // Check if a promise already exists with matching user+'|'+what
  // If not, create it
  // If so, do nothing
  let p = parseProm(urtx)
  // TODO: if the url was just alice.promises.to then we want to show alice's
  // statistics and list of promises and everything
  if (p.user === 'www' || p.user === '') { 
    // TODO: this is the case of no username, like if someone just went to
    // "promises.to" or tried to fetch "promises.to/robots.txt" or whatever
    resp.sendFile(__dirname + '/public/index.html') 
  } else if (!users.includes(p.user)) {
    // TODO: tell user "Sorry, no such user! Get in touch if you want to get in
    // on our beta!"
    resp.sendFile(__dirname + '/public/index.html')
  } else {
    Promise.findOne({ where: {urtx} })
      .then(promise => {
        if (promise) {
          console.log('duplicate', urtx);
          resp.redirect(`/promise/${urtx}`);
        } else {
          Promise.create(p);
          mailself('PROMISE', urtx);
        }
    })
 
    resp.sendFile(__dirname + '/public/index.html') 
  }
})

var listener = app.listen(process.env.PORT, function() {
  console.log('Promises app is listening on port ' + listener.address().port)
})

// --------------------------------- 80chars ---------------------------------->
